src/Virtuozzo/Helper/SocketClient.php

```diff
@@ -0,0 +1,207 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Upmind\ProvisionProviders\Servers\Virtuozzo\Helper;
+
+use SimpleXMLElement;
+use Upmind\ProvisionBase\Exception\ProvisionFunctionError;
+use Illuminate\Support\Str;
+
+class SocketClient
+{
+    /** @var resource */
+    protected $client;
+
+    protected string $host;
+
+    protected int $port;
+
+    public function __construct(string $host, int $port)
+    {
+        $this->host = $host;
+        $this->port = $port;
+    }
+
+    protected function makeAddress(): string
+    {
+        return sprintf("tcp://%s:%s", $this->host, $this->port);
+    }
+
+    public function connect(int $flags, int $timeout = 1800): void
+    {
+        if (
+            false === $client = stream_socket_client(
+                $this->makeAddress(),
+                $errorCode,
+                $errorMessage,
+                $timeout,
+                $flags,
+                stream_context_create()
+            )
+
+        ) {
+            throw ProvisionFunctionError::create("Can't connect to socket: $errorCode, $errorMessage");
+        }
+
+        if (!is_resource($client) || $errorCode !== 0) {
+            throw ProvisionFunctionError::create("Can't connect to socket: $errorCode, $errorMessage");
+        }
+
+        $this->client = $client;
+    }
+
+
+    /**
+     * @param int $length
+     *
+     * @return SimpleXMLElement
+     * @throws ProvisionFunctionError
+     */
+    public function request(string $xml): SimpleXMLElement
+    {
+        $resultXml = '';
+        $written = $this->write($xml);
+
+        if ($written > 0) {
+            //skip login response
+            $this->read();
+
+            $auth = $this->read();
+
+            $this->checkAuth($auth);
+
+            $resultXml = $this->read();
+        }
+
```

Comment:
Please implement PSR-3 logging of the request + response XMLs whenever an API call is being made, if the configuration `debug` property is enabled.